src/metabase/api/action.clj

```diff
@@ -89,7 +91,15 @@
    template               [:maybe http-action-template]
    response_handle        [:maybe json-query-schema]
    error_handle           [:maybe json-query-schema]}
+
   (api/write-check Card model_id)
+  (when (and (nil? database_id)
+             (= "query" type))
+    (throw (ex-info (tru "Must provide a database_id for query actions")
+                    {:type        type
+                     :status-code 400})))
+  (when database_id
```

Comment:
for my learning: in which case creating action without a database_id is allowed?

Reply:
`database_id` doesn't exist on the actions table itself, it's just used for query actions.

```typescript
// frontend/src/metabase/entities/actions/actions.ts L82
function createQueryAction({
  question,
  formSettings,
  ...action
}: CreateQueryActionOptions) {
  const { dataset_query, parameters, visualization_settings } =
    cleanUpQueryAction(question, formSettings);

  return ActionsApi.create({
    ...action,
    type: "query",
    dataset_query,
    database_id: dataset_query.database,
    parameters,
    visualization_settings,
  });
}
```

Reply:
@tsmacdonald I just realised this introduces a bug. Every model has a `database_id`, and all actions belong to a model. So we need to check the database_id of the model, not the database_id on the query_action. I don't know why we have a redundant database_id in the query_action table, but it's probably for performance.

Reply:
I'll update this check [here](https://github.com/metabase/metabase/issues/27751)